app/src/modules/trigger/trigger.c

```diff
@@ -9,101 +9,372 @@
 {
-	if (has_changed & button_states) {
-		trigger_work_fn(NULL);
+	int err;
+	uint8_t button_number = 1;
+
+	if (has_changed & button_states & DK_BTN1_MSK) {
+		LOG_DBG("Button 1 pressed!");
+
+		err = zbus_chan_pub(&BUTTON_CHAN, &button_number, K_SECONDS(1));
+		if (err) {
+			LOG_ERR("zbus_chan_pub, error: %d", err);
+			SEND_FATAL_ERROR();
+			return;
+		}
+	}
+}
+
+/* Convenience timer used to start/restart the RTM duration timer */
+static void rtm_duration_timer_start(void)
+{
+	LOG_DBG("Starting RTM duration timer: %d seconds", REAL_TIME_DURATION_INTERVAL_SEC);
+
+	k_timer_start(&rtm_duration_timer, K_SECONDS(REAL_TIME_DURATION_INTERVAL_SEC), K_NO_WAIT);
+}
+
+/* Zephyr State Machine framework handlers */
+
+/* HSM states:
+ *
+ * STATE_INIT: Initializing module
+ * STATE_CONNECTED: Connected to cloud. Sending triggers on timers and button presses
+ *	- STATE_REAL_TIME_MODE: Sending triggers every 10 seconds for 10 minutes
+ *	- STATE_NORMAL_MODE: Sending triggers every configured update interval
+ * STATE_DISCONNECTED: Sending of triggers are blocked
+ */
+
+/* STATE_INIT */
+
+static void init_entry(void *o)
+{
+	ARG_UNUSED(o);
+
+	LOG_DBG("init_entry");
+
+	int err = dk_buttons_init(button_handler);
+
+	if (err) {
+		LOG_ERR("dk_buttons_init, error: %d", err);
+		SEND_FATAL_ERROR();
+		return;
+	}
+}
+
+static void init_run(void *o)
+{
+	struct s_object *user_object = o;
+
+	LOG_DBG("init_run");
+
+	if ((user_object->status == CLOUD_CONNECTED) && user_object->chan == &CLOUD_CHAN) {
+		smf_set_state(SMF_CTX(&state_object), &states[STATE_CONNECTED]);
+		return;
+	}
+}
+
+/* STATE_CONNECTED */
+
+static void connected_run(void *o)
+{
+	struct s_object *user_object = o;
+
+	LOG_DBG("connected_run");
+
+	if ((user_object->status == CLOUD_DISCONNECTED) && user_object->chan == &CLOUD_CHAN) {
+		smf_set_state(SMF_CTX(&state_object), &states[STATE_DISCONNECTED]);
+		return;
+	}
+}
+
+static void connected_exit(void *o)
+{
+	ARG_UNUSED(o);
+
+	LOG_DBG("connected_exit");
+
+	LOG_DBG("Clearing timers");
+
+	k_work_cancel_delayable(&trigger_work);
+	k_timer_stop(&rtm_duration_timer);
+}
+
+/* STATE_REAL_TIME_MODE */
+
+static void real_time_mode_init(void *o)
+{
+	struct s_object *user_object = o;
+
+	LOG_DBG("ream_time_mode_init");
+
+	user_object->update_interval_current_sec = REAL_TIME_TRIGGER_INTERVAL_SEC;
+
+	rtm_duration_timer_start();
+	k_work_reschedule(&trigger_work, K_SECONDS(user_object->update_interval_current_sec));
+}
+
+static void real_time_mode_run(void *o)
+{
+	struct s_object *user_object = o;
+
+	LOG_DBG("real_time_mode_run");
+
+	if (user_object->chan == &BUTTON_CHAN) {
+		LOG_DBG("Button %d pressed in real-time mode, restarting RTM duration timer",
+			user_object->button_number);
+
+		rtm_duration_timer_start();
+	} else if (user_object->chan == &CONFIG_CHAN) {
+		LOG_DBG("Configuration received, restarting RTM duration timer");
+
+		rtm_duration_timer_start();
```

Comment:
Doesn't quite align with the state diagram here, which indicates self-transitions on `BUTTON_CHAN` and `CONFIG_CHAN` events

Reply:
This is how I intended to have it. Ill update the state diagram.